server/src/main/java/io/druid/segment/realtime/appenderator/FiniteAppenderatorDriver.java

```diff
@@ -0,0 +1,521 @@
+import java.util.Set;
+import java.util.TreeMap;
+
+/**
+ * A FiniteAppenderatorDriver drives an Appenderator to index a finite stream of data. This class does not help you
+ * index unbounded streams. All handoff is done at the end of indexing.
+ *
+ * This class helps with doing things that Appenderators don't, including deciding which segments to use (with a
+ * SegmentAllocator), publishing segments to the metadata store (with a SegmentPublisher), and monitoring handoff (with
+ * a SegmentHandoffNotifier).
+ *
+ * Note that the commit metadata stored by this class via the underlying Appenderator is not the same metadata as
+ * you pass in. It's wrapped in some extra metadata needed by the driver.
+ */
+public class FiniteAppenderatorDriver implements Closeable
+{
+  private static final Logger log = new Logger(FiniteAppenderatorDriver.class);
+
+  private final Appenderator appenderator;
+  private final SegmentAllocator segmentAllocator;
+  private final SegmentHandoffNotifier handoffNotifier;
+  private final UsedSegmentChecker usedSegmentChecker;
+  private final ObjectMapper objectMapper;
+  private final int maxRowsPerSegment;
+
+  // Key = Start of segment intervals. Value = Segment we're currently adding data to.
+  // All access to "activeSegments" and "lastSegmentId" must be synchronized on "activeSegments".
+  private final NavigableMap<Long, SegmentIdentifier> activeSegments = new TreeMap<>();
+  private volatile String lastSegmentId = null;
+
+  // Notified when segments are dropped.
+  private final Object handoffMonitor = new Object();
+
+  public FiniteAppenderatorDriver(
+      Appenderator appenderator,
+      SegmentAllocator segmentAllocator,
+      SegmentHandoffNotifierFactory handoffNotifierFactory,
+      UsedSegmentChecker usedSegmentChecker,
+      ObjectMapper objectMapper,
+      int maxRowsPerSegment
+  )
+  {
+    this.appenderator = Preconditions.checkNotNull(appenderator, "appenderator");
+    this.segmentAllocator = Preconditions.checkNotNull(segmentAllocator, "segmentAllocator");
+    this.handoffNotifier = Preconditions.checkNotNull(handoffNotifierFactory, "handoffNotifierFactory")
+                                        .createSegmentHandoffNotifier(appenderator.getDataSource());
+    this.usedSegmentChecker = Preconditions.checkNotNull(usedSegmentChecker, "usedSegmentChecker");
+    this.objectMapper = Preconditions.checkNotNull(objectMapper, "objectMapper");
+    this.maxRowsPerSegment = maxRowsPerSegment;
+  }
+
+  /**
+   * Perform any initial setup and return currently persisted commit metadata.
+   *
+   * Note that this method returns the same metadata you've passed in with your Committers, even though this class
+   * stores extra metadata on disk.
+   *
+   * @return currently persisted commit metadata
+   */
+  public Object startJob()
+  {
+    handoffNotifier.start();
+
+    final FiniteAppenderatorDriverMetadata metadata = objectMapper.convertValue(
+        appenderator.startJob(),
```

Comment:
that's ok, convertValue turns null into null, and then there's a null check in this method.